qiskit/qpy/common.py

```diff
@@ -165,6 +171,15 @@ def write_mapping(file_obj, mapping, serializer, **kwargs):
     file_obj.write(struct.pack(formats.SEQUENCE_PACK, num_elements))
     for key, datum in mapping.items():
         key_bytes = key.encode(ENCODE)
+
+        if isinstance(datum, dict):
```

Comment:
Instead of doing this, you can generate `datum_bytes` with this trick by recursively calling `write_mapping` with `container` in the code below.
https://github.com/Qiskit/qiskit-terra/blob/e9f8b7c50968501e019d0cb426676ac606eb5a10/qiskit/qpy/common.py#L196-L198